### ISSUE TITLE
[discussion] Expressing tensor dimension semantics / constraints through typing / constraints blocks. Constraints block could be scripted/traced and help for tracing/script execution and codegen

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#40373]
See https://github.com/pytorch/pytorch/issues/26889

----

It could be nice to be able to type arguments e.g. as `torch.Tensor['BCHW']` or `torch.Tensor['*C']` or `torch.LongTensor['?2']` even for user code self-documentation purposes (e.g. like one can specify dimension semantics when exporting to ONNX).

Typing information may come in different forms: maybe even some generic API for expressing frequent constraints such as "two tensors should have same first dimensions", "these two tensors should have exactly same shapes" or "these two two tensors should have broadcastable shapes".

I understand that expressing these different constraints may be a difficult task, but maybe this was already tackled somehow in functional programming community (parametrized types) and ["design by contract"](https://en.wikipedia.org/wiki/Design_by_contract) paradigm (e.g. Code Contracts integrated in C# language). Maybe some reduced / concise way that is already useful can be found

**UPD**: It would also be nice to trace/script these constraints and use them for codegen and execution to eliminate unneccessary checks and select better kernels. "expressing shaping information a bit similar to #40373. E.g. allowing tracing some constraints like assert size(X, 1) == size(Y, 2) and size(X, 2) == 64 and ndim(Z) == 3 etc and then using them for compilation/optimization (in my example in #40373 I annotate tensors with strings like BTC and BC and then check that similarly named dimensions of different tensors agree)"

cc @ezyang @bhosmer @smessmer @ljk53

**Comment by kshitij12345** [comment]
Hi,

Are you talking something along these lines? https://github.com/ofnote/tsalib

**Comment by vadimkantorov** [comment]
kind of. even if currently there is no way to enforce types, a standard "language" (or experiments towards this goal) to express them would be useful

i guess more user hints would also improve compilers / inference engines (e.g. given enough hints, it could make better use of memory allocation)

**Comment by ezyang** [comment]
Some notes on technical difficulties for doing this: I recently gave up trying to make `Module` a subclass of `Generic` in our Python files (see #38211 for the gory backstory, but in particular https://bugs.python.org/issue40897 was a show stopper) and it seems likely we would end up with a lot of similar, subtle incompatibilities if Tensor was made Generic. So of you want a parametrized Tensor type you're going to have to do it externally (e.g., through something like `python/typeshed`).

**Comment by vadimkantorov** [comment]
I think it's okay for this to be external (especially at the beginning), or it could be `torch.typing.sth` or `torchtyping.sth`.

But getting a well designed API / "language" is crucial here (even if it's only for documentation purposes). I wonder if TensorFlow Swift does anything interesting here. There must be a lot of research on this in Programmin Languages community. I think, Facebook even employs Erik Meijer who is specialist on all this. Some other properties like dtype, real/complex, layout, sparse/dense may also be interesting to include in typing sometimes.

**Comment by fylux** [comment]
Hi, I am not sure if you are aware of the current efforts toward typing tensors (e.g. https://github.com/pytorch/pytorch/issues/26889). Pyre already supports variadics that is what is needed to do things like:

```
def mm(
    left: Tensor[DType, [A, B]], right: Tensor[DType, [B, C]]
) -> Tensor[DType, [A, C]]: ...
```

You can find a more complete example [here](https://github.com/facebook/pyre-check/blob/master/examples/pytorch/sources/linear_regression.py)

We are also currently working on more advance things such as:
```
def concat(
    left: Tensor[DType, [A]], right: Tensor[DType, [B]]
) -> Tensor[DType, [A+B]]: ...
```

**Comment by ezyang** [comment]
Also relevant https://groups.google.com/forum/?oldui=1#!topic/python-shape-checkers/tczSyUG0p-4

**Comment by vadimkantorov** [comment]
Oh, cool, @apaszke is working on this :)

Shape types would already be cool and cover majority of cases. Some "Code Contracts"-like block with more complex conditions may be used for covering more complex constraints like broadcastability, promotability of types, sameness of devices, more complex type properties: real/complex/dense/sparse/some other sparse variants/device type, output padding-like shape computation,

**Comment by fylux** [comment]
Regarding actual thoughts about getting things done, here is my view of the road ahead:

1. Add support to Variadics in Python. Although this already works in Pyre there is not an official PEP yet. The most likely is that it will be written as soon as the the Parameter Specification PEP gets merged. 
2.  Probably you want to wait till most of typecheckers support it (Pyre, Mypy, Pytype)
3. Once you have it, you can proceed with typing Pytorch functions. Actually this work could be done in parallel to the PEP and just wait to be merged until Variadic support is official.
4. Repeat points 1,2,3 with the next feature that will be supported (e.g. type arithmetic).

As far as I know, @apaszke is not working on this but rather on static analysis of tensor shapes for TF Swift. 

On the other hand, regarding "Code contracts" I think that the concept is too broad so to really think about its possibilities I would suggest to try come to some examples about how and what would you express and iterate over that.

**Comment by vadimkantorov** [comment]
> On the other hand, regarding "Code contracts" I think that the concept is too broad so to really think about its possibilities I would suggest to try come to some examples about how and what would you express and iterate over that.

Agreed! Some first, frequent and actionable assertions could be: broadcastability, promotability of types, sameness of devices

**Comment by vadimkantorov** [comment]
related: https://github.com/pytorch/pytorch/pull/40493

**Comment by ezyang** [comment]
In many situations I feel like shapes and dtypes are better expressed as refinement types, rather than elevated to first class status in the "primary" types fo the functions. E.g., I want to be able to gradually shape my code, and not be obligated to have working shapes everywhere. Mypy is sort of gradually polymorphic but I'm not sure how good of a job it does at it (though this is a solved problem as far as the research literature goes.)

**Comment by vadimkantorov** [comment]
I think that if we come up with a nice API for expressing shape / dtype constraints, even without a generic static checker it would be useful. A dynamic checker (e.g. via method decorator) would already be useful and could standardize common shape checking preamble in many functions. Also, typing without checking would already be useful for documentation (including actual auto-generated documentation)

**Comment by fylux** [comment]
I agree that a formal specification of the shape constraints of each function would be useful. However, I would rather suggest to first write it down as part of the documentation (but in a formal way) and once that some API/Types exists generate the docs automatically. So basically, I would not wait till this advanced feature exists to provide users with this information, although perhaps this discussion corresponds to a separate issue.

**Comment by vadimkantorov** [comment]
How do I define a custom Tensor generic type? To support sth like:
```
import typing

Batch = typing.TypeVar('Batch')
Time = typing.TypeVar('Time')
Channels = typing.TypeVar('Channels')

# works, but isn't very nice
Tensor1 = typing.Tuple
print(Tensor1[Batch, Channels, Time])
# typing.Tuple[~Batch, ~Channels, ~Time]

class Tensor2(typing.Tuple):
  pass
print(Tensor2[Batch, Channels, Time])

#Traceback (most recent call last):
#  File "foo.py", line 13, in <module>
#    print(Tensor2[Batch, Channels, Time])
#  File "/usr/lib/python3.6/typing.py", line 682, in inner
#    return func(*args, **kwds)
#  File "/usr/lib/python3.6/typing.py", line 1253, in __getitem__
#    return super().__getitem__(parameters)
#  File "/usr/lib/python3.6/typing.py", line 682, in inner
#    return func(*args, **kwds)
#  File "/usr/lib/python3.6/typing.py", line 1131, in __getitem__
#    _check_generic(self, params)
#  File "/usr/lib/python3.6/typing.py", line 657, in _check_generic
#    raise TypeError("%s is not a generic class" % repr(cls))
#TypeError: __main__.Tensor2 is not a generic class
```

**Comment by vadimkantorov** [comment]
Is variadic support needed for this? https://github.com/python/typing/issues/193#issuecomment-479484916

**Comment by vadimkantorov** [comment]
> Hi, I am not sure if you are aware of the current efforts toward typing tensors (e.g. #26889). Pyre already supports variadics that is what is needed to do things like: https://github.com/facebook/pyre-check/blob/master/examples/pytorch/sources/linear_regression.py

Oh, this is very cool!

**Comment by ezyang** [comment]
Support for variadics in mypy out-of-the-box is *really* bad. You'll almost definitely need to do a mypy plugin, or maybe use pyre instead.

**Comment by vadimkantorov** [comment]
I'd first like to do simply annotation for doc purposes, without even any type checking, so this should not be an issue...

**Comment by ezyang** [comment]
Still fairly certain that variadics are a "soup for me, not for you!" mypy feature 👍

**Comment by vadimkantorov** [comment]
Aliasing `Tensor = typing.Tuple` works :)

**Comment by vadimkantorov** [comment]
My simple "shapeguard"-like experiment: https://gist.github.com/vadimkantorov/71155662378b46cd07b6287f6be7c951

**Comment by vadimkantorov** [comment]
Comment originally at https://github.com/pytorch/pytorch/pull/45828#issuecomment-715539223:

"expressing shaping information a bit similar to #40373. E.g. allowing tracing some constraints like assert size(X, 1) == size(Y, 2) and size(X, 2) == 64 and ndim(Z) == 3 etc and then using them for compilation/optimization (in my example in #40373 I annotate tensors with strings like BTC and BC and then check that similarly named dimensions of different tensors agree)"

**Comment by mrahtz** [comment]
Heads up that we're working on a draft of a PEP for variadic generics in [PEP 646](https://www.python.org/dev/peps/pep-0646/). If you're interested, give it a read and let us know any feedback in [this thread](https://mail.python.org/archives/list/typing-sig@python.org/thread/WFV5K2LK3LFQSO63X2KUOCK3VVLAQ374/) in typing-sig. Thanks!

**Comment by vadimkantorov** [comment]
This is a super-long thread :) I'll read it and if I have anything to add - will post there.

A remark about my current efforts at poor-person's crude shape typing:
1) sometimes it makes sense to create "anonymous" dimensions, or express `...` as a variable number of dimensions, unimportant for the current code
2) sometimes it makes sense to have just Height, Width, but sometimes adding extra info/tags to them Height_Input, Height_Output
3) sometimes it is useful to express that a number of dimensions have been flattened e.g. [Height, Width, Channels] -> [Height x Width, Channels]
4) sometimes it is very readable to have some super-concise dimension name which is a number, e.g. 3 or 2 for representing channel dimension of rgb images or 2 for stereo-audio. unfortunately Python identifiers can't start with a digit

**Comment by ezyang** [comment]
Since the conversation on this might be heating up soon, I'm going to close this as a dupe of #26889 to make sure everyone sees the same discussion

**Comment by ezyang** [comment]
Reopening this issue just to discuss typing via assert statements

**Comment by vadimkantorov** [comment]
Maybe this could drive design of the "typing constraints" (both static and dynamic) or "compiler hints", their representations, how to access them etc, how compilers/optimizers could use them for optimization (e.g. memory register planning for inference).

**Comment by vadimkantorov** [comment]
@ezyang btw `assert x.shape[1] == 3` statements can be used for decaying dynamic shape dims into fixed shape dims
